src/power_grid_model/data_types.py

```diff
@@ -13,7 +13,7 @@
 
 from power_grid_model.core.dataset_definitions import ComponentType
 
-ComponentTypeVar = TypeVar("ComponentTypeVar", ComponentType, str)
+ComponentTypeLike = TypeVar("ComponentTypeLike", bound=ComponentType | str)
```

Comment:
Technically breaking, but I don't think any of our users uses this as of now
I can also do
```suggestion
ComponentTypeLike = TypeVar("ComponentTypeLike", bound=ComponentType | str)
ComponentTypeVar = ComponentTypeVar
```

Reply:
Why does it become necessary now to use `bound`? Also your can-also-do has a typo.

Reply:
If `ComponentTypeVar` was removed altogether, why not just name this `ComponentTypeVar` instead. That shouldn't be breaking, right?

Reply:
@Jerry-Jinfeng-Guo I believe `bound` must be together with `|` to indicate that it can not only be `ComponentType` or `str`, but also subsets of them and subsets of their `union`. If one was to remove bound, I am confident that Pylint, or some other linter, would complain if the user inputs `ComponentType`s with both `ComponentType` and `str`, and that is why initially (before Python 3.10) we had to explicitly write all the type combinations, which made it really ugly and long.

Reply:
> If ComponentTypeVar was removed altogether, why not just name this ComponentTypeVar instead. That shouldn't be breaking, right?

I originally had

```suggestion
ComponentTypeVar = TypeVar("ComponentTypeVar", bound=ComponentType | str)
```

(your suggestion) but i have a problem with how it's used in the code base, not with the object. However, renaming would be maybe breaking if the users are already using it.

>> Why does it become necessary now to use bound?
>
> I believe bound must be together with | to indicate that it can not only be ComponentType or str, but also subsets of them and subsets of their union

@figueroa1395 is right: `bounds` enables subsets of types. Strict matching would prevent mixed combinations and actually also makes deductions from subsets trickier. `bounds` can do both in one go

> Also your can-also-do has a typo.

whoops, thanks

Reply:
> (your suggestion) but i have a problem with how it's used in the code base, not with the object. However, renaming would be maybe breaking if the users are already using it.

So I understand that the problem is how we, as devs., use it internally, correct? Hence why this naming convention suits it best within the codebase itself?

If so, doing `ComponentTypeVar: TypeAlias = ComponentTypeLike`, as you suggested, should keep it clean within the codebase and prevent breaking changes for the users (I would say let's assume they already use it, just to be safe).

Reply:
I believe the problem comes from the fact that the user could do things like:
```
line = initialize_array(DatasetType.input, ComponentType.line, 1)
...

input_data = {
    "line": line,
    ...
}
```
or
```
input_data = {
    ComponentType.node: node,
    "line": line,
}
```
Hence mixing `literals` with the new `ComponenType`s. Such mixing should be supported, but linters usually hate it, even though it should be straight forward. Introducing `TypeVar` with `bound` will make sure no linter complains and that users can freely mix notation (They shouldn't, but it is probably going to happen).

Reply:
FYI: i applied above suggestion. if agree with it, please resolve the comment and approve. otherwise, let's have a quick call

Reply:
I approved changes. If @Jerry-Jinfeng-Guo is okay with these answers, please resolve.

Reply:
Agree to disagree. Please proceed without my agreement.

Reply:
Potential compromise after continuing on #663

Reply:
cfr. offline sync: this is the solution we go with

```py
ComponentTypeLike = ComponentType | str
ComponentTypeVar = TypeVar("ComponentTypeVar", bound=ComponentTypeLike)  # helper used for type deduction
```